### PR TITLE
Refactor EXPR classes.

### PR DESCRIPTION
**Pull request by JonHanna**
Convert GetXXX and SetXXX methods to auto-properties.

Convert public fields to auto-properties.

Rename to be PascalCase

Restricted to cases that are individually simple.

### CODE REVIEW

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Conversion.cs [hunk 1]
```diff
@@ -351,48 +351,48 @@ public bool canConvert(CType src, CType dest)
...
             {
                 // Conversion works.
-                checkUnsafe(expr.type); // added to the binder so we don't bind to pointer ops
+                checkUnsafe(expr.Type); // added to the binder so we don't bind to pointer ops
                 checkUnsafe(dest); // added to the binder so we don't bind to pointer ops
                 return exprResult;
             }
 
-            if (expr.isOK() && !dest.IsErrorType())
+            if (expr.IsOK&& !dest.IsErrorType())
```
Comment (stephentoub) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Conversion.cs]: Nit: spacing

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/GroupToArgsBinder.cs [hunk 2]
```diff
@@ -1356,13 +1356,13 @@ private void ReportErrorsOnFailure()
                     // if we tried to bind to an extensionmethod and the instance argument Type Inference failed then the method does not exist
                     // on the type at all. this is treated as a lookup error
                     CType type = null;
-                    if (_pGroup.GetOptionalObject() != null)
+                    if (_pGroup.OptionalObject!= null)
```
Comment (stephentoub) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/GroupToArgsBinder.cs]: Nit: spacing

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/GroupToArgsBinder.cs [hunk 3]
```diff
@@ -1356,13 +1356,13 @@ private void ReportErrorsOnFailure()
...
                     {
-                        type = _pGroup.GetOptionalObject().type;
+                        type = _pGroup.OptionalObject.Type;
                     }
-                    else if (_pGroup.GetOptionalLHS() != null)
+                    else if (_pGroup.OptionalLHS!= null)
```
Comment (stephentoub) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/GroupToArgsBinder.cs]: Nit: spacing (will stop commenting on these)

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Types/Type.cs [hunk 4]
```diff
@@ -198,8 +198,18 @@ private static Type CalculateAssociatedSystemTypeForAggregate(AggregateType aggt
...
+        public bool IsType
+        {
+            get
+            { return true; }
+        }
+
+        public bool IsNamespace
+        {
+            get
+            { return false; }
```
Comment (stephentoub) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Types/Type.cs]: Nit: these newlines are unnecessary.  They could also be done with expression-bodied members.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/AssemblyQualifiedNamespaceSymbol.cs [hunk 5]
```diff
@@ -18,14 +18,20 @@ internal sealed class AssemblyQualifiedNamespaceSymbol : ParentSymbol, ITypeOrNa
         // AssemblyQualifiedNamespaceSymbol
         // ----------------------------------------------------------------------------
 
-        public bool IsType()
+        public bool IsType
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/AssemblyQualifiedNamespaceSymbol.cs]: nit: we can also use the Syntax:
public bool IsType => false;
for concise representation.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs [hunk 6]
```diff
@@ -46,19 +46,19 @@ protected override EXPR Dispatch(EXPR expr)
 
         /////////////////////////////////////////////////////////////////////////////////
         // Statement types.
-        protected override EXPR VisitASSIGNMENT(EXPRASSIGNMENT assignment)
+        protected override Expr VisitASSIGNMENT(ExprAssignment assignment)
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs]: >VisitASSIGNMENT [](start = 32, length = 15)

VisitAssignment?

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs [hunk 7]
```diff
@@ -69,56 +69,56 @@ protected override EXPR VisitASSIGNMENT(EXPRASSIGNMENT assignment)
...
+            Expr rhs = Visit(pExpr.Operator);
+            Expr lhs = Visit(pExpr.Left);
             return GenerateCall(PREDEFMETH.PM_EXPRESSION_ASSIGN, lhs, rhs);
         }
 
         /////////////////////////////////////////////////////////////////////////////////
         // Expression types.
 
-        protected override EXPR VisitBOUNDLAMBDA(EXPRBOUNDLAMBDA anonmeth)
+        protected override Expr VisitBOUNDLAMBDA(ExprBoundLambda anonmeth)
```
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs]: >VisitBOUNDLAMBDA [](start = 32, length = 16)

Same comment in rest of methods, if it belongs to the same PR.

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs (hunk 6 above)
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Visitors/ExpressionTreeRewriter.cs]: There are more similar changes to be done, but I think at 2000+ lines just from doing Expr an types derived from it I'll leave others for later.

**Comment by stephentoub** [comment]
@JonHanna, thanks for working on this.  In the future, please submit updates based on CR feedback on a PR like this as a separate commit, so we can just re-review what changed rather than re-reviewing all changes across 77 files :smile: We can then 'squash and merge' to consolidate it to a single commit when merging.

**Pull request merged**